Reverse the ownership between SwiftASTContextForExpressions and TypeS…

adrian-prantl opened the pull request:
…ystemSwiftTypeRef(ForExpressions)

This is - believe it or not - an NFC commit.

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwift.h:
```diff
@@ -69,11 +69,10 @@ class TypePayloadSwift {
-///                              ↓    ↓
-///      SwiftASTContextForModules    SwiftASTContextForExpressions
-///          (deprecated)
+///    TypeSystemSwiftTypeRef ⟷ SwiftASTContextFormModules (deprecated)
```
```suggestion
///    TypeSystemSwiftTypeRef ⟷ SwiftASTContextForModule (deprecated)
```

kastiglione commented on lldb/source/Target/Target.cpp:
```diff
@@ -2546,10 +2553,11 @@ llvm::Optional<SwiftASTContextReader> Target::GetScratchSwiftASTContext(
     bool fallback = true;
-    auto typesystem_sp = SwiftASTContext::CreateInstance(
-        lldb::eLanguageTypeSwift, *lldb_module, nullptr, this, fallback);
+    auto typesystem_sp = std::shared_ptr<TypeSystemSwiftTypeRefForExpressions>(
```
make_shared?

kastiglione commented on lldb/source/Target/Target.cpp:
```diff
@@ -2546,10 +2553,11 @@ llvm::Optional<SwiftASTContextReader> Target::GetScratchSwiftASTContext(
+                                                 *this, *lldb_module));
     auto *swift_ast_ctx =
-        llvm::cast<SwiftASTContextForExpressions>(typesystem_sp.get());
+        llvm::cast<TypeSystemSwiftTypeRefForExpressions>(typesystem_sp.get());
```
Is this cast a no-op?

kastiglione commented on lldb/source/Target/Target.cpp:
```diff
@@ -2546,10 +2553,11 @@ llvm::Optional<SwiftASTContextReader> Target::GetScratchSwiftASTContext(
       DisplayFallbackSwiftContextErrors(global_scratch_ctx);
 
     bool fallback = true;
```
this is unused

adrian-prantl commented:
I'm now down to zero test failures locally.

adrian-prantl commented:
@swift-ci test

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:
```diff
@@ -783,6 +784,8 @@ TypeSystemSwiftTypeRef::GetSwiftified(swift::Demangle::Demangler &dem,
   if (ident.empty())
     return node;
 
+  if (ident.equals("NSURL") || ident.equals("URL"))
+    llvm::errs();
```
not sure if you meant to commit this (mentioning in case this causes CI failures)

adrian-prantl committed:
Reverse the ownership between SwiftASTContextForExpressions

and TypeSystemSwiftTypeRef(ForExpressions).

his is an NFC commit. This patch introduces a new class
TypeSystemSwiftTypeRefForExpressions which is being returned by
Target::getSwiftScratchTypeSystem(). It owns a
SwiftASTContextForExpressions that is initialized lazily when
GetSwiftASTContext() is called.

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
Finally got this to pass after factoring out the (even more) NFC parts. The problem was the order in which the destructors of the various SwiftASTContext classes were deleting themselves from the global ASTMap.

adrian-prantl commented:
@swift-ci test
